WalletWasabi/ExchangeRate/ExchangeRateProvider.cs

```diff
@@ -0,0 +1,70 @@
+using System.Collections.Immutable;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using ExchangeRateProviderInfo = (string Name, bool ClearNetOnly, string ApiUrl, System.Func<string, decimal> Extractor);
+
+namespace WalletWasabi.ExchangeRate;
+
+public class ExchangeRateProvider(EndPoint? socksProxyEndPoint = null)
+{
+	public static readonly ImmutableArray<ExchangeRateProviderInfo> Providers =
+	[
+		("bitstamp", true, "https://www.bitstamp.net/api/v2/ticker/btcusd", JsonPath(".bid")),
+		("blockchain", false, "https://blockchain.info/ticker", JsonPath(".USD.buy")),
+		("coinbase", true, "https://api.coinbase.com/v2/exchange-rates?currency=BTC", JsonPath(".data.rates.USD")),
+		("coingecko", false, "https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=bitcoin", JsonPath(".[0].current_price")),
+		("coingate", true, "https://api.coingate.com/v2/rates/merchant/BTC/USD", JsonPath("$")),
+		("gemini", false, "https://api.gemini.com/v1/pubticker/btcusd", JsonPath(".bid"))
+	];
+
+	public async Task<ExchangeRate> GetExchangeRateAsync(string providerName, CancellationToken cancellationToken)
+	{
+		var providerInfo = Providers.FirstOrDefault(x => x.Name.Equals(providerName, StringComparison.InvariantCultureIgnoreCase));
+		if (providerInfo == default)
+		{
+			throw new NotSupportedException($"Exchange rate provider '{providerName}' is not supported.");
+		}
+		var proxy = GetWebProxy();
+		if (providerInfo.ClearNetOnly && proxy is not null)
+		{
+			throw new NotSupportedException($"Exchange rate provider '{providerName}' cannot be reached with Tor, only clearnet.");
+		}
+
+		var url = new Uri(providerInfo.ApiUrl);
+
+#pragma warning disable RS0030 // Do not use banned APIs
+		using var httpClientHandler = new HttpClientHandler();
+		using var httpClient = new HttpClient(httpClientHandler);
+		httpClientHandler.Proxy = proxy;
+		httpClient.BaseAddress = new Uri($"{url.Scheme}://{url.Host}");
+		httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("WasabiWallet", Helpers.Constants.ClientVersion.ToString()));
+#pragma warning restore RS0030 // Do not use banned APIs
+
+		using var response = await httpClient.GetAsync(url.PathAndQuery, cancellationToken).ConfigureAwait(false);
+		var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+		var rate = providerInfo.Extractor(json);
+		return new ExchangeRate("USD", rate);
+	}
+
+	private WebProxy? GetWebProxy()
+	{
+		return socksProxyEndPoint switch
```

Comment:
Why not use `WasabiHttpClientFactory.NewHttpClient`? Implementing this new WebProxy logic should be part of the `ExchangeRateProvider.cs`?

Reply:
Q1:

nopara73 created the socks5 client for interacting with Tor for one very good reason: it was the only way to use Tor as a socks5 proxy. On top of that he built an TorHttpClient from scratch, this is, a component that understands the HTTP protocol. Then Kimi build another layer of machinery on top including pools, connections and other stuff. 

NicolasDorier also had to implement the socks5 proxy but he did it as a real proxy. Why did he implemented that? because there wasn't any other way, if you wanted to interact with Tor you had to code your own socks5 logic.

Nicolas' design is better because he didn't have to implement his own HttpClient, instead he used the `System.Net.Http.HttpClient` and configured it to use his own proxy.

However, a couple of years ago, in .NET 6 Microsoft implemented full support for socks5 proxies and then we can interact with Tor without needing to code absolutely anything. This means that we can use a much better and feature-complete `HttpClient` instead of having to code one ourselves, this means that we can use websocket without needing to implement the protocol by ourselves, this means that al compression/decompression and all versions of Http are supported without us to have to implement more and more complex Http protocols be ourselves.

So, one day we have to replace all our custom implementation code (because we don't need it anymore) and use the (HttpClient, WebProxy) dupla. I would prefer to stop using the old code asap.

Q2:
No, but this is the only class that requires it and until I find a better place for it I think it is okay where it is not (but I will move it somewhere else in the near future)

Reply:
> I would prefer to stop using the old code asap.

Thank you for the elaboration. I acknowledged the new direction. 



> (but I will move it somewhere else in the near future)

ACK.